client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt

```diff
@@ -0,0 +1,187 @@
+ * threads, sleeping, or backoff.
+ *
+ * This is not thread safe.
+ */
+internal class EmbeddedBackfillRun<B : Backfill<*, *>>(
+  private val operator: BackfillOperator<*, *>,
+  override val dryRun: Boolean,
+  override val parameters: Map<String, ByteString>,
+  override val rangeStart: String?,
+  override val rangeEnd: String?,
+
+  override var batchSize: Long = 100L,
+  override var scanSize: Long = 10_000L,
+  override var computeCountLimit: Long = 1L
+) : BackfillRun<B> {
+  override val backfill: B
+    get() = operator.backfill as B
+  override val prepareBackfillResponse: PrepareBackfillResponse
+
+  private val precomputeProgress: Map<String, MutableInstanceCursor>
+  override var precomputeMatchingCount: Long = 0L
+  override var precomputeScannedCount: Long = 0L
+
+  private val scanProgress: Map<String, MutableInstanceCursor>
+  override val instanceProgressSnapshot: Map<String, InstanceCursor>
+    get() = scanProgress.mapValues { it.value.snapshot() }
+
+  private val batchesToRun: ArrayDeque<BatchSnapshot> = ArrayDeque()
+  override val batchesToRunSnapshot: List<BatchSnapshot>
+    get() = batchesToRun.toList()
+
+  /** Prepares a backfill for a run. */
+  init {
+    prepareBackfillResponse = operator.prepareBackfill(PrepareBackfillRequest.Builder()
+        .backfill_name(operator.name())
+        .range(KeyRange(rangeStart?.encodeUtf8(), rangeEnd?.encodeUtf8()))
+        .parameters(parameters)
+        .dry_run(dryRun)
+        .build())
+    precomputeProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
+          it.instance_name, it.backfill_range)
+    }
+    scanProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
+          it.instance_name, it.backfill_range)
+    }
+  }
+
+  override fun precomputeScan(): GetNextBatchRangeResponse {
+    val notDone = precomputeProgress.filterNot { it.value.done }
+    check(notDone.isNotEmpty()) { "Nothing left to precompute" }
+    val cursor = notDone.values.first()
+    val response =
+        operator.getNextBatchRange(GetNextBatchRangeRequest.Builder()
+            .instance_name(cursor.instanceName)
+            .backfill_range(cursor.keyRange)
+            .previous_end_key(cursor.previousEndKey)
+            .parameters(parameters)
+            .batch_size(batchSize)
+            .scan_size(scanSize)
+            .compute_count_limit(computeCountLimit)
+            .precomputing(true)
+            .dry_run(dryRun)
+            .build())
+    if (response.batches.isEmpty()) {
+      cursor.done = true
+    } else {
+      cursor.previousEndKey = response.batches.map { it.batch_range.end }.max()
+      response.batches.onEach {
+        precomputeMatchingCount += it.matching_record_count
+        precomputeScannedCount += it.scanned_record_count
+      }
+    }
+    return response
+  }
+
+  override fun precomputeRemaining() {
+    do {
```

Comment:
aww yeah do while

Reply:
I can put that all on one line too. Very easy to understand. Kotlin is great.

Reply:
The autoformatter doesn't like it though. So I'm leaving it alone.